app/com/linkedin/drelephant/tuning/engine/SparkHBTParamRecommender.java

```diff
@@ -431,61 +344,32 @@ private int getMemoryIncreaseForGC() {
-      }
-      if (lastRunDynamicAllocationMaxExecutors != null) {
-        suggestedDynamicAllocationMaxExecutors =
-            (int) Math.ceil(lastRunDynamicAllocationMaxExecutors * lastRunExecutorCore / suggestedCore);
-      }
-      if (lastRunExecutorInstances != null) {
-        suggestedExecutorInstances = (int) Math.ceil(lastRunExecutorInstances * lastRunExecutorCore / suggestedCore);
-      }
-      validateNumExecutorParameters();
+      suggestExecutorInstancesBasedOnCore();
     } else {
-      suggestedDynamicAllocationMinExecutors = lastRunDynamicAllocationMinExecutors;
-      suggestedDynamicAllocationMaxExecutors = lastRunDynamicAllocationMaxExecutors;
       suggestedExecutorInstances = lastRunExecutorInstances;
     }
+    if (suggestedExecutorInstances > CLUSTER_DEFAULT_DYNAMIC_ALLOCATION_MAX_EXECUTOR) {
+      suggestedExecutorInstances = CLUSTER_DEFAULT_DYNAMIC_ALLOCATION_MAX_EXECUTOR;
+    }
+    //job fails if max executors is less than executor instances
+    if (lastRunDynamicAllocationMaxExecutors != null
+        && suggestedExecutorInstances > lastRunDynamicAllocationMaxExecutors) {
+      suggestedExecutorInstances = lastRunDynamicAllocationMaxExecutors;
+    }
   }
 
-  /**
-   * Validating executor parameters
-   * Anything should not be less than 1 and more than MAX_EXECUTORS
-   */
-  private void validateNumExecutorParameters() {
-    if (suggestedDynamicAllocationMinExecutors != null) {
-      if (suggestedDynamicAllocationMinExecutors < 1) {
-        suggestedDynamicAllocationMinExecutors = 1;
-      }
-      if (suggestedDynamicAllocationMinExecutors > MAX_EXECUTORS) {
-        suggestedDynamicAllocationMinExecutors = MAX_EXECUTORS;
+  private void suggestExecutorInstancesBasedOnCore() {
+    Integer lastRunActualExecutorInstances = lastRunExecutorInstances;
+    if (lastRunDynamicAllocationEnabled && lastRunDynamicAllocationMaxExecutors != null) {
+      if (lastRunActualExecutorInstances == null) {
+        lastRunActualExecutorInstances = lastRunDynamicAllocationMaxExecutors;
       }
     }
-    if (suggestedDynamicAllocationMaxExecutors != null) {
-      if (suggestedDynamicAllocationMinExecutors != null
-          && suggestedDynamicAllocationMaxExecutors < suggestedDynamicAllocationMinExecutors) {
-        suggestedDynamicAllocationMaxExecutors = suggestedDynamicAllocationMinExecutors;
-      }
-      if (suggestedDynamicAllocationMaxExecutors > MAX_EXECUTORS) {
-        suggestedDynamicAllocationMaxExecutors = MAX_EXECUTORS;
-      }
-    }
-    if (suggestedExecutorInstances != null) {
-      if (suggestedExecutorInstances < 1) {
-        suggestedExecutorInstances = 1;
-      }
-      if (suggestedExecutorInstances > MAX_EXECUTORS) {
-        suggestedExecutorInstances = MAX_EXECUTORS;
-      }
+    if (lastRunActualExecutorInstances != null) {
+      suggestedExecutorInstances =
+          (int) Math.ceil(lastRunActualExecutorInstances.doubleValue() * lastRunExecutorCore / suggestedCore);
```

Comment:
Do u want to add some buffer number of Instances , just to be on safer side (in terms of run time)

Reply:
I think if number of available cores remain same, there should not be much impact of run time. We will test this on few jobs and then may be define if we need some buffer.

Reply:
ok more core may cause more GC (for compute intensive jobs) ,which may cause increase in run time . But as you have said , test on few jobs and then may add buffer